File: tomcat/webapps/MassBankDevelKit/script/DbManager.js
/*******************************************************************************
 *
 * Copyright (C) 2014 MassBank Project
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *******************************************************************************
 *
 * DbManager.js
 *
 * ver 1.0.0 Feb. 3, 2014
 *
 ******************************************************************************/
var hostName
var ipAddress;
var inputUrlHtml;

/**
 *
 */
function initLoad(hostName, ipAddress) {
	this.hostName = hostName;
	this.ipAddress = ipAddress;
	this.inputUrlHtml = document.getElementById("inputUrl").innerHTML;
	if ( !!document.formEdit ) {
		var objForm = document.formEdit;
		var siteNo = objForm.siteNo.value;
		objForm.siteNo.focus();
		setBgColor();
		if ( siteNo == -1 ) {
			objForm.siteDb.readOnly = false;
			objForm.siteDb.className = "";
			objForm.siteUrl.readOnly = true;
			objForm.siteUrl.className = "readOnly";
			objForm.btnAdd.disabled = false;
			objForm.btnEdit.disabled = true;
			objForm.btnDelete.disabled = true;
			document.formEdit.siteUrl.value = getMyServerUrl();
		}
		else if ( siteNo == 0 ) {
			objForm.siteDb.readOnly = true;
			objForm.siteDb.className = "readOnly";
			objForm.btnAdd.disabled = true;
			objForm.btnEdit.disabled = false;
			objForm.btnDelete.disabled = true;
			showInternalUrlSelect();
		}
		else {
			var tagid = "no" + siteNo;
			if ( document.getElementById(tagid + "Type" ).value == "internal" ) {
				objForm.siteDb.readOnly = true;
				objForm.siteDb.className = "readOnly";
				objForm.siteUrl.readOnly = true;
				objForm.siteUrl.className = "readOnly";
			}
			objForm.btnAdd.disabled = true;
			objForm.btnEdit.disabled = false;
			objForm.btnDelete.disabled = false;
		}
	}
}

/**
 *
 */
function selectNo() {
	var objForm = document.formEdit;
	var siteNo = objForm.siteNo.value;
	var siteNoList = document.formEdit.siteNo.options;
	setBgColor();
	showInputUrl();

	if ( siteNo == -1 ) {
		objForm.siteDb.value = "";
		objForm.siteShortLabel.value = "";
		objForm.siteLongLabel.value = "";
		objForm.siteType[0].checked = true;
		objForm.siteType[0].disabled = false;
		objForm.siteType[1].disabled = false;
		objForm.siteType[0].checked = true;
		setClass('intLabel', 'readOnly', '');
		setClass('extLabel', 'readOnly', '');
		objForm.siteUrl.readOnly = true;
		objForm.siteUrl.className = "readOnly";
		objForm.siteDb.readOnly = false;
		objForm.siteDb.className = "";
		objForm.btnAdd.disabled = false;
		objForm.btnEdit.disabled = true;
		objForm.btnDelete.disabled = true;
		document.formEdit.siteUrl.value = getMyServerUrl();
	}
	else {
		var tagid = "no" + siteNo;
		var element1 =  document.getElementById(tagid + "Db");
		var element2 =  document.getElementById(tagid + "ShortLabel");
		var element3 =  document.getElementById(tagid + "LongLabel");
		var element4 =  document.getElementById(tagid + "Url");
		if ( navigator.userAgent.indexOf("Firefox") == -1 ) {
			db = element1.innerText;
			sLabel = element2.innerText;
			lLabel = element3.innerText;
			val4 = element4.innerText;
		}
		else {
			db = element1.textContent;
			sLabel = element2.textContent;
			lLabel = element3.textContent;
			url = element4.textContent;
		}
		objForm.siteDb.value = db;
		objForm.siteShortLabel.value = sLabel;
		objForm.siteLongLabel.value = lLabel;
		objForm.siteUrl.value = url;

		if ( document.getElementById(tagid + "Type").value == "internal" ) {
			objForm.siteDb.readOnly = true;
			objForm.siteDb.className = "readOnly";
			objForm.siteType[0].disabled = false;
			objForm.siteType[1].disabled = true;
			objForm.siteType[0].checked = true;
			setClass('intLabel', 'readOnly', '');
			setClass('extLabel', '', 'readOnly');
			objForm.siteUrl.readOnly = true;
			objForm.siteUrl.className = "readOnly";
		}
		else {
			objForm.siteDb.readOnly = false;
			objForm.siteDb.className = "";
			objForm.siteType[0].disabled = true;
			objForm.siteType[1].disabled = false;
			objForm.siteType[1].checked = true;
			setClass('intLabel', '', 'readOnly');
			setClass('extLabel', 'readOnly', '');
			objForm.siteUrl.readOnly = false;
			objForm.siteUrl.className = "";
		}
		if ( siteNo == 0 ) {
			objForm.siteUrl.readOnly = false;
			objForm.siteUrl.className = "";
			objForm.btnAdd.disabled = true;
			objForm.btnEdit.disabled = false;
			objForm.btnDelete.disabled = true;
			showInternalUrlSelect();
		}
		else {
			objForm.btnAdd.disabled = true;
			objForm.btnEdit.disabled = false;
			objForm.btnDelete.disabled = false;
		}
	}
}

/**
 *
 */
function selectUrlType() {
	objForm = document.formEdit;
	var siteNo =  document.formEdit.siteNo.value;
	if ( siteNo == -1 ) {
		if ( objForm.siteType[1].checked ) {
			objForm.siteUrl.readOnly = false;
			objForm.siteUrl.className = "";
			objForm.siteUrl.value = "http://";
		}
		else {
			objForm.siteUrl.readOnly = true;
			objForm.siteUrl.className = "readOnly";
			document.formEdit.siteUrl.value = getMyServerUrl();
		}
	}
}

/**
 *
 */
function setClass(elementId, className1, className2) {
	if ( !op && !ie && !ns4 && !ns6 ) {
		alert("Your browser is not supported.");
		return;
	}
	if ( ns4 ) {//NS4
		element = document.layers[elementId];
	}
	else {		//OP,IE,NS6
		element = document.getElementById(elementId);
	}
	if ( element.className == className1 ) {
		element.className = className2;
	}
}

/**
 *
 */
function setBgColor() {
	var OFF_COLOR = "WhiteSmoke";
	var ON_COLOR = "LightSteelBlue";
	var objForm = document.formEdit;
	var siteNo = objForm.siteNo.value;
	var siteNoList = document.formEdit.siteNo.options;
	var num = siteNoList.length;
	if ( num == 1 ) {
		return;
	}
	for ( i = 1; i < num; i++ ) {
		var id = "row" + siteNoList[i].value;
		var val = OFF_COLOR;
		if ( siteNo != -1 && siteNoList[i].value == siteNo ) {
			val = ON_COLOR;
		}
		document.getElementById(id).style.background = val;
	}
}

/**
 *
 */
function getMyServerUrl() {
	var element = document.getElementById("no0Url");
	if ( navigator.userAgent.indexOf("Firefox") == -1 ) {
		val = element.innerText;
	} else {
		val = element.textContent;
	}
	return val;
}

/**
 *
 */
function checkInputValue(act) {
	var objForm = document.formEdit;
	var siteNo = objForm.siteNo.value;
	var siteUrl = objForm.siteUrl.value;
	var url = siteUrl.toLowerCase();
	var isInternal = isInternalUrl(url);
	if ( objForm.siteType[0].checked && !isInternal ) {
		alert("Please enter a URL of internal");
		return false;
	}
	else if ( objForm.siteType[1].checked && isInternal ) {
		alert("Please enter a URL of external");
		return false;
	}
	objForm.act.value = act;
	if ( act == "delete" ) {
		return confirm("Are you sure?");
	}
	return true;
}

/**
 *
 */
function isInternalUrl(url) {
	if (   url.indexOf("localhost") != -1
		|| url.indexOf("127.0.0.1") != -1 
		|| (this.hostName != "" && url.indexOf(hostName) != -1 )
		|| (this.ipAddress != "" && url.indexOf(ipAddress) != -1) ) {
		return true;
	}
	return false;
}

/**
 *
 */
function showInternalUrlSelect() {
	var html =  "<select name=\"siteUrl\" style=\"width:100%;\">\n";
	var names = [this.hostName, this.ipAddress, "localhost"];
	for ( var i = 0; i < names.length; i++) {
		var url = "http://" + names[i] + "/MassBank/";
		var selected = "";
		if ( url == getMyServerUrl() ) {
			selected = " selected";
		}
		html += "<option value=\"" + url + "\"" + selected + ">" + url + "</option>";
	}
	html +=  "</select>\n";
	var element = document.getElementById("inputUrl");
	element.innerHTML = html;
}

/**
 *
 */
function showInputUrl() {
	var element = document.getElementById("inputUrl");
	element.innerHTML = this.inputUrlHtml;
}

File: tomcat/webapps/MassBankDevelKit/src/net/massbank/admin/DbManager.java
/*******************************************************************************
 *
 * Copyright (C) 2014 MassBank Project
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *******************************************************************************
 *
 * DbManager.java
 *
 * ver 1.0.0 Feb. 3, 2014
 *
 ******************************************************************************/
package net.massbank.admin;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Set;
import javax.servlet.jsp.JspWriter;
import net.massbank.core.common.GetConfig;
import org.apache.commons.lang.BooleanUtils;
import org.apache.commons.lang.NumberUtils;
import org.apache.commons.lang.StringUtils;

public class DbManager {
	public static final String NO = "NO";
	public static final String DB_NAME = "DB Name";
	public static final String SHORT_LABEL = "Short Label";
	public static final String LONG_LABEL = "Long Label";
	public static final String URL = "URL";
	public static final String URL_TYPE = "URL Type";
	private AdminInfo adminInfo = new AdminInfo();
	private HashMap<String, String> inputs = null;
	private int inputNo = 0;
	private String inputDbName  = "";
	private String inputSlabel  = "";
	private String inputLlabel  = "";
	private String inputUrl     = "";
	private boolean inputType = false;
	private SetConfig setConf = null;
	private JspWriter out = null;

	/**
	 * constructor
	 */
	public DbManager(HashMap<String, String> inputs, String rootPath, JspWriter out) {
		this.inputs = inputs;
		this.inputNo     = NumberUtils.stringToInt(inputs.get("siteNo"));
		this.inputDbName = inputs.get("siteDb");
		this.inputSlabel = inputs.get("siteShortLabel");
		this.inputLlabel = inputs.get("siteLongLabel");
		this.inputUrl    = inputs.get("siteUrl");
		this.inputType   = BooleanUtils.toBoolean(inputs.get("siteType"));
		this.setConf = new SetConfig(rootPath);
		this.out = out;
	}

	/**
	 * add
	 */
	public boolean add() {
		if ( !checkInputValue(true) ) {
			return false;
		}
		String sql = "create database " + this.inputDbName;
		if ( !AdminDbUtil.updateDB("", sql) ) {
			return false;
		}
		GetConfig conf = new GetConfig();
		String[] confDbNames = conf.getDbNames();
		String[] tableNames = { "PEAK", "RECORD", "COMPOUND_INFO", "COMPOUND_NAMES", "INSTRUMENT" };
		for ( String tableName : tableNames ) {
			sql = "create table " + this.inputDbName
				+ "." + tableName+ " like " + confDbNames[0] + "." + tableName;
			if ( !AdminDbUtil.updateDB("", sql) ) {
				return false;
			}
		}
		return this.setConf.add(this.inputSlabel, this.inputLlabel, this.inputUrl, this.inputDbName);
	}


	/*
	 * edit
	 */
	public boolean edit() {
		if ( !checkInputValue(false) ) {
			return false;
		}
		ArrayList<Integer> internalNoList = new ArrayList();
		int[] indexes = null;
		if ( this.inputNo == 0 ) {
			GetConfig conf = new GetConfig();
			String[] urls = conf.getSiteUrls();
			String serverUrl = urls[0];
			for ( int i = 0; i < urls.length; i++ ) {
				if ( serverUrl.equals(urls[i]) ) {
					internalNoList.add(i);
				}
			}
			indexes = new int[internalNoList.size()];
			for ( int i = 0; i < internalNoList.size(); i++ ) {
				indexes[i] = internalNoList.get(i);
			}
		}
		else {
			indexes = new int[]{this.inputNo};
		}
		return this.setConf.update(indexes, this.inputSlabel, this.inputLlabel, this.inputUrl);
	}

	/**
	 * delete
	 */
	public boolean delete() {
		GetConfig conf = new GetConfig();
		String[] dbNames = conf.getDbNames();
		String sql = "drop database " + dbNames[this.inputNo];
		if ( !AdminDbUtil.updateDB("", sql) ) {
			return false;
		}
		return this.setConf.delete(this.inputNo);
	}


	/*
	 * get informations
	 */
	public String[] getInformations() {
		return this.adminInfo.getList();
	}

	/**
	 * check input values
	 */
	private boolean checkInputValue(boolean isCheckDB) {
		boolean ret = true;
		String msg = "";
		Collection vals = this.inputs.values();
		Set keySet = this.inputs.keySet();
		Iterator ite = keySet.iterator();
		while ( ite.hasNext() ) {
			Object key = ite.next();
			String val = (String)this.inputs.get(key);
			if ( StringUtils.isBlank(val) ) {
				adminInfo.add(AdminInfo.INFO_LEVEL_ERROR, msg, "");
				ret = false;
			}
			else if ( !StringUtils.isAsciiPrintable(val) ) {
				adminInfo.add(AdminInfo.INFO_LEVEL_ERROR, msg, "" );
				ret = false;
			}
		}

		boolean isUrl = true;
		try {
			new URL(this.inputUrl);
			if ( this.inputUrl.replaceAll("http://", "").replaceAll("https://", "").trim().equals("") ) {
				isUrl = false;
			}
		}
		catch ( MalformedURLException e) {
			isUrl = false;
		}
		if ( !isUrl ) {
			try {
				out.println( "<font color=\"red\"><b>The URL is not correct.</b></font><br>" );
			}
			catch (Exception e) {
				e.printStackTrace();
			}
			ret = false;
		}

		if ( isCheckDB ) {
			String[] chars = { "\\", "/", ":", "*", "?", "\"", "<", ">", "|", "." };
			if ( StringUtils.indexOfAny(this.inputDbName, chars) != -1 ) {
				msg = DB_NAME + "contains illegal characters";
				adminInfo.add(AdminInfo.INFO_LEVEL_ERROR, msg, "" );
				ret = false;
			}
			else if ( AdminDbUtil.existDB(this.inputDbName) ) {
				try {
					out.println( "<font color=\"red\"><b>Can't create database '" + this.inputDbName + "' as it already exists.</b></font><br>" );
				}
				catch (Exception e) {
					e.printStackTrace();
				}
				ret = false;
			}
		}
		return ret;
	}
}
